Libs/MRML/Core/vtkMRMLNode.cxx

```diff
@@ -506,41 +514,140 @@ void vtkMRMLNode::ParseReferencesAttribute(const char* attValue, std::set<std::s
-              int index = this->GetNumberOfNodeReferences(role.c_str()) - 1;
-              this->SetNthNodeReferenceProperty(role.c_str(), index, propertyKey.c_str(), propertyValue.c_str());
+              std::size_t propertySeparator = property.find_first_of('=');
+              if (propertySeparator != std::string::npos)
+              {
+                std::string propertyKey = property.substr(0, propertySeparator);
+                propertyKey = this->NodeReferencePropertyDecodeString(propertyKey);
+                std::string propertyValue = property.substr(propertySeparator + 1);
+                propertyValue = this->NodeReferencePropertyDecodeString(propertyValue);
+                properties[propertyKey] = propertyValue;
+              }
             }
           }
         }
         else
         {
           // Current value string is the node ID for the current node reference.
-          this->AddNodeReferenceID(role.c_str(), value.c_str());
-          references.insert(role);
+          roleReferences.push_back(std::make_pair(value, std::map<std::string, std::string>()));
         }
       }
+      references.insert(role);
     }
     start = (end == std::string::npos) ? std::string::npos : end + 1;
     end = attribute.find_first_of(';', start);
     sep = attribute.find_first_of(':', start);
   }
+
+  // Now update the actual node references to match the new references
+  // Use a surgical approach: only modify references that need to be changed
+  for (const auto& newRoleEntry : newReferences)
+  {
+    const std::string& role = newRoleEntry.first;
+    const std::vector<std::pair<std::string, std::map<std::string, std::string>>>& newRoleReferences = newRoleEntry.second;
+
+    // Get current references for this role
+    std::vector<const char*> currentNodeIDs;
+    this->GetNodeReferenceIDs(role.c_str(), currentNodeIDs);
+
+    // Build list of new node IDs for comparison
+    std::vector<std::string> newNodeIDs;
+    for (const auto& refPair : newRoleReferences)
+    {
+      newNodeIDs.push_back(refPair.first);
+    }
+
+    // Remove references to node IDs that are no longer needed first
+    // This reduces the number of existing items that might need updates
+    for (int currentIndex = static_cast<int>(currentNodeIDs.size()) - 1; currentIndex >= 0; currentIndex--)
+    {
+      const char* currentNodeID = currentNodeIDs[currentIndex];
+      if (currentNodeID)
+      {
+        // Check if this node ID is still needed in the new reference list
+        bool nodeIdStillNeeded = false;
+        for (const std::string& newNodeID : newNodeIDs)
+        {
+          if (std::string(currentNodeID) == newNodeID)
+          {
+            nodeIdStillNeeded = true;
+            break;
+          }
+        }
+
+        // If this node ID is no longer needed, remove the reference
+        if (!nodeIdStillNeeded)
+        {
+          this->RemoveNthNodeReferenceID(role.c_str(), currentIndex);
+        }
+      }
+    }
+
```

Comment:
```suggestion

    // Re-fetch after removals to avoid stale indice
    currentNodeIDs.clear();
    this->GetNodeReferenceIDs(role.c_str(), currentNodeIDs);
```

Reply:
@lassoan 
Since we call `RemoveNthNodeReferenceID`,should we get the updated list IDs ?

Reply:
Good point, I'll add this.